### DISCUSSION TITLE
ISR on demand(api/revalidate) works locally but it fails on VERCEL

### DISCUSSION THREAD

**Discussion by diegovicente5** [vercel/next.js#71949] [Help] [answered]
### Summary

I deployed my website to Vercel, but before that I validated all features in my local server including the api/revalidate and it works as expected, however when I deployed my to prod the revalidate did not work. The error that I got in the vercel logs is "Failed to revalidate /: Invalid response 500"

Note. I have a api proxy in order to call external api(`/pages/api/[...slug].ts`) from next and also `pages/api/revalidate.ts`. The api proxy is used only for client api call, and from the server side I am just using the direct domain to the external api.

### Additional information

```js
{
"message": "Error revalidating",
"error": "Failed to revalidate /: Invalid response 500"
}
```

**Comment by icyJoseph** [comment]
Do you have any more logs? Does the external API actually receive the requests you proxy to it? Are you perhaps missing an environmental variable?

**Reply by diegovicente5** [reply]
The implementation works locally in dev and prod mode. I added all env variables in Vercel and even the build work fine but the problem is only when I try to access to api/revalidate/ page. I tested the url even passing a different secret code and it return the error that the secret code is wrong but it breaks in this line `await res.revalidate(path);`. 


Please look, this is my current api revalidate logic. 
apps/gudyz/gudyz_ec/src/pages/api/revalidate.ts

```
import { NextApiRequest, NextApiResponse } from 'next';

export const dynamic = 'force-dynamic';
export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse
) {
  const secret = process.env.VERCEL_REVALIDATE_TOKEN;
  const path = req.query.path as string;

  if (req.query.secret !== secret) {
    return res.status(401).json({ message: 'Invalid secret token.' });
  }
  res.setHeader('Cache-Control', 'no-store');
  // res.setHeader('x-vercel-cache', 'MISS');
  try {
    await res.revalidate(path);
    return res.json({ revalidated: true });
  } catch (err: any) {
    console.error('Revalidation error:', err);
    return res
      .status(500)
      .json({ message: 'Error revalidating', error: err.message });
  }
}
```

and that's what I see in the log
<img width="1435" alt="Screenshot 2024-10-28 at 10 37 51 AM" src="https://github.com/user-attachments/assets/a50a01cf-a8c1-4d7e-ac36-e1a7fe210931">
s.

**Reply by icyJoseph** [reply]
Yeah, so the problem is not the revalidation handler, I think, but that when your page reruns to revalidate, something crashes. 

The Vercel logs you have there, probably contain hints at what's going on.

You do have the same Node.js version locally and on Vercel right? What does the revalidated page do? And again, do you have any logs on that external service (proxied to)?

**Reply by diegovicente5** [reply]
I am using nodejs 20 in my LS and in Vercel.

Basically the revalidate should get the latest content of the API.

I have an API proxy only for the FE api call, for the API calls called from the server side I am not using the api proxy, I am just using the direct domain of the external API.

In LS it works fine.
<img width="1439" alt="Screenshot 2024-10-28 at 10 58 34 AM" src="https://github.com/user-attachments/assets/b3526c19-7f87-44b8-9cfc-b9a7358bcad9">

**Reply by icyJoseph** [reply]
Time to introduce some more logging on that server to external API flow... Let's see where it drops

**Reply by diegovicente5** [reply, marked as answer]
the issue was resolved. Looks like the package-lock.json has some missing dependencies and that was breaking the revalidate. I deleted the file package-lock.json and sent to rebuild on Vercel, now the revalidate is working as expected.

**Reply by icyJoseph** [reply]
Uff tricky situation. Might have been some dependencies that used to be devDependencies only? And then changed but not updated on the lock file. Weird, but great that you could debug it that far! Sorry for the LLM/ChatGpt noise you had to put up with as well.

**Comment by aniruddhaadak80** [comment]
Hi @diegovicente5 ,
---

You're getting a 500 error when using `api/revalidate` on Vercel, even though it works locally.

### Potential Solutions
1. **Check API Response**: Ensure the external API is returning valid responses.
2. **Environment Variables**: Verify that all environment variables are correctly set on Vercel.
3. **Examine Logs**: Look at the Vercel logs for detailed error messages.
4. **Network Access**: Make sure the external API is accessible from Vercel's servers.

### Error Example
```json
{
  "message": "Error revalidating",
  "error": "Failed to revalidate /: Invalid response 500"
}
```

Review these aspects to troubleshoot the revalidation issue.

**Reply by diegovicente5** [reply]
Hi @aniruddhaadak80 ,
I added more details of the issue in this [comment](https://github.com/vercel/next.js/discussions/71949#discussioncomment-11076435), Also regarding your potential solutions, the build works fine and the ISR pages are loaded which means that the api response and environment variables are fine. Regarding the network access, I believe that it is fine because the build works fine, and finally regarging the vercel logs, after every deploy I got some different error like this:

`Revalidation error: {
  message: 'Failed to revalidate /products/mascotas: Invalid response 500',
  stack: 'Error: Failed to revalidate /products/mascotas: Invalid response 500\n' +
    '    at z (/var/task/node_modules/next/dist/compiled/next-server/pages-api.runtime.prod.js:20:11895)\n' +
    '    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)\n' +
    '    at async d (/var/task/dist/apps/gudyz/gudyz_ec/.next/server/pages/api/revalidate.js:1:718)\n' +
    '    at async K (/var/task/node_modules/next/dist/compiled/next-server/pages-api.runtime.prod.js:20:16881)\n' +
    '    at async U.render (/var/task/node_modules/next/dist/compiled/next-server/pages-api.runtime.prod.js:20:17520)\n' +
    '    at async r6.runApi (/var/task/node_modules/next/dist/compiled/next-server/server.runtime.prod.js:17:44381)\n' +
    '    at async r6.handleCatchallRenderRequest (/var/task/node_modules/next/dist/compiled/next-server/server.runtime.prod.js:17:38387)\n' +
    '    at async r6.runImpl (/var/task/node_modules/next/dist/compiled/next-server/server.runtime.prod.js:16:17376)\n' +
    '    at async r6.handleRequestImpl (/var/task/node_modules/next/dist/compiled/next-server/server.runtime.prod.js:16:16465)\n' +
    '    at async Server.<anonymous> (/var/task/apps/gudyz/gudyz_ec/___next_launcher.cjs:26:5)',
  code: undefined
}`

**Reply by aniruddhaadak80** [reply]
Hi @vijayth2-cerebras,

Thanks for the update! The `500` response during revalidation indicates an issue with the API endpoint. Here are a few steps to troubleshoot:

1. **Check API Endpoint**: Ensure it's functioning correctly and returning the expected response.
2. **Add Logging**: Include logging in your API route to catch any errors.
3. **Verify Environment Variables**: Confirm that all required environment variables are set correctly in Vercel.
4. **Review Vercel Logs**: Check the logs for any additional clues about the error.
